### ISSUE TITLE
[RFC #1] Drop kango framework in favor of chrome.* API

### ISSUE THREAD

**Issue by iamolivinius** [secure-software-engineering/rose#183]
Due to the ongoing incompatibility between firefox and kango framework, the unresponsive maintainer, and the fact that extensions build with kango are rejected from the firefox extension store I suggest to drop the usage of kango framework in future ROSE development.

In future releases Firefox will support several `chrome.*` APIs, see: https://developer.mozilla.org/en-US/Add-ons/WebExtensions 
In other words we could write code for chrome, opera, and firefox without the need of a cross platform framework. 

Unfortunately, supporting safari won't be possible anymore.

Referring to http://gs.statcounter.com/#desktop-browser-ww-monthly-201409-201509-bar this way we would be able to cover 72,35% of all browsers world wide. Sticking to kango would mean we could cover 57,6%.

**Comment by dharma-guardian** [comment]
I support it. In a next version of ROSE, we should switch to this approach. Considering the amount of time we needed to make certain frameworks work with kangoAPI in order to implement ROSEv3, it makes total sense to break free from this dependency, when larger parts of ROSE should be changed.

**Comment by apoller** [comment]
How much effort will it be to also support IE and Safari? Especially for Mac Users I don't know how common Safari is. IE might not be a big issue as it is mostly used in corporate networks only.

**Comment by iamolivinius** [comment]
I can't find much information about add-on development for IE but it looks like add-ons for IE need to be compiled. Microsoft Edge does not support add-ons/extensions at all right now: https://dev.modern.ie/platform/faq/extensions

Support for safari is doable. Safari supports a different API but this could be handled by a abstraction layer.

**Comment by dharma-guardian** [comment]
Before getting on with 3.1 we should check this for feasibility.

**Comment by dharma-guardian** [comment]
If we should go with it, I would throw [typescript](https://www.typescriptlang.org/) into the ring to drop babel.

**Comment by iamolivinius** [comment]
Nice idea but would break support for safari completely. Async / await can
be to transpiled to es6 only and generator functions are not supported by
safari yet.

**Comment by dharma-guardian** [comment]
Didn't realize that. But Safari Technology Preview already supports it: http://kangax.github.io/compat-table/es6/#test-generators
So, let's see, when we get 3.1 started...

**Comment by iamolivinius** [comment]
If https://github.com/secure-software-engineering/rose/issues/183#issuecomment-219369793 is a serious proposal it should be moved into it's own RFC.

In my opinion TS will become interesting as soon as it hits version 2.0. Until then there is no async/await transformation and no strict null checking. Additionally we should discuss why it is a good idea to replace one compiler for another one.

**Comment by eppfel** [comment]
https://github.com/EmailThis/extension-boilerplate

**Comment by ohadbeck** [comment]
I am also going through the same process of removing the kango framework dependency from a extension I am working on. I was wondering if you guys have any insight or tips on how to do this. I am stuck and the only resource I found was this thread. I am specifically stuck on trying to change the fabfile so it does not go through kango and to generate the extension. Any help or insight would be greatly appreciated as I am very stuck! 
Thanks!

**Comment by eppfel** [comment]
No, experience, because we never tried. But the abstraction layer would be the way to go I guess.

An example for that I stumbled ago was Signal Desktop, which was a Chrome extension. They only implemented chrome and not firefox, because of the switch to electron. But it is an abstraction layer anyway: https://github.com/signalapp/Signal-Desktop/blob/chrome/js/chromium.js

The rest would be just to generate the manifest file and folder structure for each browser. If you don#t expect to much changes in the accord, you could do this manually.

**Comment by ohadbeck** [comment]
I see. Thank you so much! that helps me a ton

**Comment by eppfel** [comment]
Make sure to share your results, if you something done 😉
